### PR TITLE
[REF] *: gantt unavailabilities

### PR DESCRIPTION
**Pull request by Polymorphe57**
We adapt the code of mrp to changes brought by https://github.com/odoo/enterprise/pull/64263. Since the gantt view for mrp workorders always needs to have unavailabilities for some workcenters in order to compute the aggregates displayed in the row "Total", we have overridden get_gantt_data here to compute those unavailabilities. This has allowed us to make the mrp_workorder_gantt view much simpler.

### CODE REVIEW

File: addons/mrp/models/mrp_workorder.py [hunk 1]
```diff
@@ -552,38 +552,18 @@ def _cal_cost(self):
...
-                workcenter_ids.add(row.get('resId'))
-
-        for row in rows:
-            traverse_inplace(search_workcenter_ids, row)
-        start_datetime = fields.Datetime.to_datetime(start_date)
-        end_datetime = fields.Datetime.to_datetime(end_date)
-        workcenters = self.env['mrp.workcenter'].browse(workcenter_ids)
-        unavailability_mapping = workcenters._get_unavailability_intervals(start_datetime, end_datetime)
+        if field_name != 'workcenter_id':
+            super().gantt_unavailability(start_datetime, end_datetime, scale, field_name, field_values)
```
Comment (Polymorphe57) [inline addons/mrp/models/mrp_workorder.py:558]: do we really need to call super or return {}? What is the good practice?

File: addons/mrp/models/mrp_workorder.py [hunk 2]
```diff
@@ -552,38 +552,35 @@ def _cal_cost(self):
...
+            workcenter_ids = set()
+            if groupby and 'workcenter_id' in groupby:
+                for group in gantt_data['groups']:
+                    res_id = group['workcenter_id'][0] if group['workcenter_id'] else False
+                    workcenter_ids.add(res_id)
+            else:
+                for record in gantt_data['records']:
+                    res_id = record['workcenter_id']['id'] if record.get('workcenter_id') else False
+                    workcenter_ids.add(res_id)
+            unavailabilities = self.gantt_unavailability(['workcenter_id'], {'workcenter_id': workcenter_ids}, start_date, end_date, scale)
```
Comment (Polymorphe57) [inline addons/mrp/models/mrp_workorder.py:567]: do we need gantt_unavailability and _gantt_unavailability?

File: addons/mrp/models/mrp_workorder.py [hunk 3]
```diff
@@ -552,38 +552,35 @@ def _cal_cost(self):
         return total
 
     @api.model
-    def gantt_unavailability(self, start_date, end_date, scale, group_bys=None, rows=None):
+    def get_gantt_data(self, domain, groupby, read_specification, limit=None, offset=0, unavailability_fields=[], start_date=None, end_date=None, scale=None):
+        gantt_data = super().get_gantt_data(domain, groupby, read_specification, limit=limit, offset=offset, unavailability_fields=unavailability_fields, start_date=start_date, end_date=end_date, scale=scale)
+        if not gantt_data['unavailabilities'].get('workcenter_id'):
```
Comment (Polymorphe57) [inline addons/mrp/models/mrp_workorder.py:557]: what about case gantt_data['unavailabilities'] = {} ?

File: addons/mrp/models/mrp_workorder.py [hunk 4]
```diff
@@ -552,38 +552,35 @@ def _cal_cost(self):
...
-                workcenter_ids.add(row.get('resId'))
-
-        for row in rows:
-            traverse_inplace(search_workcenter_ids, row)
-        start_datetime = fields.Datetime.to_datetime(start_date)
-        end_datetime = fields.Datetime.to_datetime(end_date)
-        workcenters = self.env['mrp.workcenter'].browse(workcenter_ids)
-        unavailability_mapping = workcenters._get_unavailability_intervals(start_datetime, end_datetime)
+        if field_name != 'workcenter_id':
+            return super()._gantt_unavailability(start_datetime, end_datetime, scale, field_name, field_values)
```
Comment (Polymorphe57) [inline addons/mrp/models/mrp_workorder.py:575]: call super or return {}?

File: addons/mrp/models/mrp_workorder.py [hunk 5]
```diff
@@ -552,38 +552,35 @@ def _cal_cost(self):
...
 
-        def add_unavailability(row, workcenter_id=None):
-            if row.get('groupedBy') and row.get('groupedBy')[0] == 'workcenter_id' and row.get('resId'):
-                workcenter_id = row.get('resId')
-            if workcenter_id:
-                row['unavailabilities'] = [{'start': interval[0], 'stop': interval[1]} for interval in unavailability_mapping[workcenter_id]]
-                return {'workcenter_id': workcenter_id}
+        workcenters = self.env['mrp.workcenter'].browse(field_values)
```
Comment (Polymorphe57) [inline addons/mrp/models/mrp_workorder.py:577]: what if field_values = [False]? what if field_values = [1, False]?

File: addons/mrp/models/mrp_workorder.py (hunk 3 above)
Comment (Polymorphe57) [reply addons/mrp/models/mrp_workorder.py:557]: if 'workcenter_id' not in gantt_data['unavailabilities']

**Review by aab-odoo — APPROVED** [review]
robodoo r+

**Pull request closed without merging**